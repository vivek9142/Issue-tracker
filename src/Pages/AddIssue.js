import React from 'react';
import {Formik,Form} from 'formik';
import IssueField from '../Components/IssueField';
import {Prompt} from 'react-router-dom'; 
import {useDispatch} from 'react-redux';
import { addIssue } from '../Redux/Actions/IssueActions';

import * as Yup from 'yup';

const AddIssue = (props) =>{

    const dispatch = useDispatch();
    let formIsHalfFilled ;

    const onSubmit = (event) => {
        const newIssue = {};
        newIssue.views=0;
        newIssue.description = event.description;
        newIssue.severity = event.severity;
        newIssue.status = event.status;
        newIssue.createdDate = event.createdDate;
        newIssue.resolvedDate = event.resolvedDate;
        dispatch(addIssue(newIssue)).then((res)=>{
            // if(res.type.indexOf('fulfilled') !== -1) 
                
        });
    }

    const validate = Yup.object({
        description:Yup.string().required('Issue Description is required'),
        
        severity:Yup.string()
        .oneOf(['Critical','Minor','Major'],`Severity should be 'Critical','Minor','Major'`)
        .required('Severity is required'),
        
        status:Yup.string()
        .required('Status is required'),

        createdDate:Yup.date()
        .required('createdDate is required'),

        resolvedDate:Yup.date()
        .required('resolvedDate is required')
    });
    
    return(
    <React.Fragment>
        <Prompt when={formIsHalfFilled} message="You have unsaved changes,sure You want to leave?"/>
        
        <h1>Add Issue</h1>

        <Formik initialValues={
            {
                description:'',
                severity:'',
                status:'',
                createdDate:'',
                resolvedDate:''
            } }
            validationSchema = {validate}
            onSubmit={async (values)=> onSubmit(values)}
        >
           {formik=>{
               let values = Object.values(formik.values);
               formIsHalfFilled = values.filter(val => {return val && val !== ''}).length>0;
               return(
               <div>
                   <Form>
                        <IssueField label='Description' name='description' type='text'/>
                        <IssueField label='Severity' name='severity' type='select'/>
                        <IssueField label='Status' name='status' type='radio-button'/>
                        <IssueField label='Created Date' name='createdDate' type='date'/>
                        <IssueField label='Resolved Date' name='resolvedDate' type='date'/>
                        <button style={{marginLeft:`1rem`}} type='submit' >Submit</button>
                   </Form>
               </div>
           )}} 
        </Formik>
    </React.Fragment>
    )
};

export default AddIssue;