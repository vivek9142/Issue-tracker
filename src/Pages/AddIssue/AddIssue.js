import React from 'react';
import {Formik,Form} from 'formik';
import IssueField from '../../Components/IssueField/IssueField';
import {Prompt} from 'react-router-dom'; 
import {useDispatch} from 'react-redux';
import { addIssue } from '../../Redux/Actions/IssueActions';

import * as Yup from 'yup';

const AddIssue = (props) =>{
    let formIsHalfFilled;

    const dispatch = useDispatch();
    

    const onSubmit = (event) => {
        const newIssue = {};
        newIssue.views=0;
        newIssue.description = event.description;
        newIssue.severity = event.severity;
        newIssue.status = event.status;
        newIssue.createdDate = event.createdDate;
        newIssue.resolvedDate = event.resolvedDate;
        newIssue.id=Math.floor(Math.random()*1000);
        dispatch(addIssue(newIssue)).then((res)=>{
            props.history.push('/');  
        });
    }

    const validate = Yup.object({
        description:Yup.string().required('Issue Description is required'),
        
        severity:Yup.string()
        .oneOf(['Critical','Minor','Major'],`Severity should be 'Critical','Minor','Major'`)
        .required('Severity is required'),
        
        status:Yup.string()
        .required('Status is required'),

        createdDate:Yup.date()
        .required('createdDate is required'),

        resolvedDate:Yup.date()
        .required('resolvedDate is required')
    });

    return(
    <React.Fragment>

        
        
        <div className="container user-section mt-5 mb-3 ">
            <h3 className="text-center">
                Add Issue
            </h3>
            <hr/>
        </div>

        <Formik initialValues={
            {
                description:'',
                severity:'',
                status:'',
                createdDate:'',
                resolvedDate:''
            } }
            validationSchema = {validate}
            onSubmit={async (values)=> onSubmit(values)}
        >
           {formik=>{
               let values = Object.values(formik.values);
               formIsHalfFilled = values.filter(val => {return val && val !== ''});
               formIsHalfFilled = formIsHalfFilled.length>0 && formIsHalfFilled.length<5;
               
               return(
                   <>
                    <Prompt when={formIsHalfFilled} message="You have unsaved changes,sure You want to leave?"/>
                    <div className="container user-section--auth-container">
                            <Form>
                                <div className="row g-3">
                                        <div className="col-12">
                                            <IssueField label='Description' name='description' type='text'/>
                                        </div>
                                        <div className="col-md-6 col-sm-12">
                                            <IssueField label='Severity' name='severity' type='select'/>
                                        </div>
                                        <div className="col-md-6 col-sm-12">
                                            <IssueField label='Status' name='status' type='radio-button'/>
                                        </div>
                                        <div className="col-md-6 col-sm-12">
                                            <IssueField label='Created Date' name='createdDate' type='date'/>
                                        </div>
                                        <div className="col-md-6 col-sm-12">
                                            <IssueField label='Resolved Date' name='resolvedDate' type='date'/>
                                        </div>

                                        <div className="col-12" style={{'marginBottom':'5rem'}}>
                                            <button type="submit" className="btn btn-primary">Submit</button>
                                        </div>
                                </div>
                            </Form>
                    </div>
                </>
           )}} 
        </Formik>
    </React.Fragment>
    )
};

export default AddIssue;