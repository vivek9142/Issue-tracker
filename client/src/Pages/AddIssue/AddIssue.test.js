import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import AddIssue from './AddIssue';
import {Provider} from 'react-redux';
import store from '../../Redux/Store';
import { MemoryRouter } from 'react-router-dom';

describe('AddIssue Page Test Suite', () => {
    beforeEach(()=>{
        
    render(
        <MemoryRouter>   
        <Provider store={store}><AddIssue/></Provider>
        </MemoryRouter> 
        );
    });
    test('AddIssue Snapshot',()=>{
        const tree = renderer.create(<MemoryRouter>   
            <Provider store={store}><AddIssue/></Provider>
            </MemoryRouter> );
        expect(tree).toMatchSnapshot();
      });
      
    test('renders main Heading of AddIssue Page', () => {
        const element = screen.getByText(/Add Issue/i);
        expect(element).toBeInTheDocument();
    });

    test('AddIssue should have inputs', () => {
        expect(screen.getAllByRole('textbox')).toHaveLength(1);
        expect(screen.getAllByRole('combobox')).toHaveLength(1);
        expect(screen.getAllByRole('group')).toHaveLength(1);
    });

    test('AddIssue should have a submit button', () => {
        const submitButton = screen.getByRole('button', { name: 'Submit' });
        expect(submitButton).toBeDefined()
    });

  
})