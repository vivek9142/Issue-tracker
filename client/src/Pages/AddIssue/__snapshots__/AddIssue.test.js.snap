// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddIssue Page Test Suite AddIssue Snapshot 1`] = `
Array [
  <div
    className="container user-section mt-5 mb-3 "
  >
    <h3
      className="text-center"
    >
      Add Issue
    </h3>
    <hr />
  </div>,
  <div
    className="container user-section--auth-container"
  >
    <form
      action="#"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="row g-3"
      >
        <div
          className="col-12"
        >
          <div
            style={
              Object {
                "padding": "1rem",
              }
            }
          >
            <label
              className="form-label"
              htmlFor="Description"
            >
              Description
            </label>
            <input
              className="form-control"
              name="description"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
            <div
              style={
                Object {
                  "height": "1rem",
                }
              }
            />
          </div>
        </div>
        <div
          className="col-md-6 col-sm-12"
        >
          <div
            style={
              Object {
                "padding": "1rem",
              }
            }
          >
            <label
              className="form-label"
              htmlFor="severity"
            >
              Severity
            </label>
            <select
              className="form-select"
              name="severity"
              onBlur={[Function]}
              onChange={[Function]}
              value=""
            >
              <option
                value="Select Option"
              >
                Select Option
              </option>
              <option
                value="Minor"
              >
                Minor
              </option>
              <option
                value="Major"
              >
                Major
              </option>
              <option
                value="Critical"
              >
                Critical
              </option>
            </select>
            <div
              style={
                Object {
                  "height": "1rem",
                }
              }
            />
          </div>
        </div>
        <div
          className="col-md-6 col-sm-12"
        >
          <div
            style={
              Object {
                "padding": "1rem",
              }
            }
          >
            <label
              htmlFor="status"
            >
              Status
            </label>
            <div
              aria-labelledby="my-radio-button"
              className="form-check"
              role="group"
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <div
                className="form-check form-check-inline"
              >
                <input
                  checked={false}
                  className="form-check-input"
                  name="status"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="radio"
                  value="Open"
                />
                <label
                  className="form-check-label"
                  htmlFor="Open"
                >
                  Open
                </label>
              </div>
              <div
                className="form-check form-check-inline"
              >
                <input
                  checked={false}
                  className="form-check-input"
                  name="status"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="radio"
                  value="In Progress"
                />
                <label
                  className="form-check-label"
                  htmlFor="In Progress"
                >
                  In Progress
                </label>
              </div>
              <div
                className="form-check form-check-inline"
              >
                <input
                  checked={false}
                  className="form-check-input"
                  name="status"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="radio"
                  value="Closed"
                />
                <label
                  className="form-check-label"
                  htmlFor="Closed"
                >
                  Closed
                </label>
              </div>
              <div
                style={
                  Object {
                    "height": "1rem",
                  }
                }
              />
            </div>
          </div>
        </div>
        <div
          className="col-md-6 col-sm-12"
        >
          <label
            htmlFor="Created Date"
          >
            Created Date
          </label>
          <input
            className="form-control"
            name="createdDate"
            onBlur={[Function]}
            onChange={[Function]}
            type="date"
            value=""
          />
          <div
            style={
              Object {
                "height": "1rem",
              }
            }
          />
        </div>
        <div
          className="col-md-6 col-sm-12"
        >
          <label
            htmlFor="Resolved Date"
          >
            Resolved Date
          </label>
          <input
            className="form-control"
            name="resolvedDate"
            onBlur={[Function]}
            onChange={[Function]}
            type="date"
            value=""
          />
          <div
            style={
              Object {
                "height": "1rem",
              }
            }
          />
        </div>
        <div
          className="col-12"
          style={
            Object {
              "marginBottom": "5rem",
            }
          }
        >
          <button
            className="btn btn-primary"
            type="submit"
          >
            Submit
          </button>
        </div>
      </div>
    </form>
  </div>,
]
`;
